core/src/main/scala/ai/lum/odinson/state/EventPromoter.scala

```diff
@@ -0,0 +1,102 @@
+package ai.lum.odinson.state
+
+import ai.lum.odinson.EventMatch
+import ai.lum.odinson.NamedCapture
+import ai.lum.odinson.OdinsonMatch
+import ai.lum.odinson.StateMatch
+import ai.lum.odinson.lucene.OdinResults
+import ai.lum.odinson.lucene.search.OdinsonScoreDoc
+
+class EventPromoter {
+
+  protected def countChildren(odinsonMatch: OdinsonMatch): Int = {
+    val count = odinsonMatch.namedCaptures.foldLeft(0) { case (total, namedCapture) =>
+      total + 1 + countChildren(namedCapture.capturedMatch)
+    }
+
+    count
+  }
+
+  protected def newLabeledNamedOdinResults(namedCapture: NamedCapture, odinsonScoreDoc: OdinsonScoreDoc): LabeledNamedOdinResults = {
+    val odinsonMatch = namedCapture.capturedMatch
+    val childCount = countChildren(odinsonMatch)
+
+    LabeledNamedOdinResults(
+      namedCapture.label,
+      Some(namedCapture.name),
+      new OdinResults(
+        1 + childCount,
+        Array(
+          new OdinsonScoreDoc(
+            doc = odinsonScoreDoc.doc,
+            score = Float.NaN,
+            shardIndex = odinsonScoreDoc.shardIndex,
+            matches = Array(namedCapture.capturedMatch),
+            segmentDocId = odinsonScoreDoc.segmentDocId,
+            segmentDocBase = odinsonScoreDoc.segmentDocBase
+          )
+        )
+      )
+    )
+  }
+
+  protected def promoteOdinsonMatch(odinsonMatch: OdinsonMatch, odinsonScoreDoc: OdinsonScoreDoc): Array[LabeledNamedOdinResults] = {
+
+    def promoteOdinsonMatch(odinsonMatch: OdinsonMatch): Array[LabeledNamedOdinResults] = {
+      val result = odinsonMatch match {
+        // This is only ever considered in the case of an EventMatch.
+        case eventMatch: EventMatch =>
+          val names = eventMatch.namedCaptures.map(_.name)
+          // It is required that there are no duplicate names in the namedCaptures.
+          // This is not true!
+          // require(names.distinct.size == eventMatch.namedCaptures.length)
+          // Because of this, the namedCapture can be stored more than once in the state.
+          val nameToPromote = eventMatch.argumentMetadata
```

Comment:
@kwalcock -- i haven't fully traced this, so IDK, but this seems to be slightly overly complicated for the task...?
Was there some situation you were trying to handle that led to this, rather than a more straighforward loop with a push to the state?

Reply:
See also #186.  Yes, it is overly complicated.